components/brave_rewards/browser/test/rewards_p3a_browsertest.cc

```diff
@@ -0,0 +1,222 @@
+
+namespace rewards_browsertest {
+
+class RewardsP3ABrowserTest : public InProcessBrowserTest,
+                              public brave_rewards::RewardsServiceObserver {
+ public:
+  RewardsP3ABrowserTest() {
+    contribution_ = std::make_unique<RewardsBrowserTestContribution>();
+    promotion_ = std::make_unique<RewardsBrowserTestPromotion>();
+    response_ = std::make_unique<RewardsBrowserTestResponse>();
+    histogram_tester_.reset(new base::HistogramTester);
+  }
+
+  void SetUpOnMainThread() override {
+    InProcessBrowserTest::SetUpOnMainThread();
+
+    context_helper_ =
+        std::make_unique<RewardsBrowserTestContextHelper>(browser());
+
+    // HTTP resolver
+    host_resolver()->AddRule("*", "127.0.0.1");
+    https_server_.reset(new net::EmbeddedTestServer(
+        net::test_server::EmbeddedTestServer::TYPE_HTTPS));
+    https_server_->SetSSLConfig(net::EmbeddedTestServer::CERT_OK);
+    https_server_->RegisterRequestHandler(
+        base::BindRepeating(&rewards_browsertest_util::HandleRequest));
+    ASSERT_TRUE(https_server_->Start());
+
+    // Rewards service
+    brave::RegisterPathProvider();
+    auto* profile = browser()->profile();
+    rewards_service_ = static_cast<brave_rewards::RewardsServiceImpl*>(
+        brave_rewards::RewardsServiceFactory::GetForProfile(profile));
+    rewards_service_->AddObserver(this);
+
+    // Response mock
+    base::ScopedAllowBlockingForTesting allow_blocking;
+    response_->LoadMocks();
+    rewards_service_->ForTestingSetTestResponseCallback(base::BindRepeating(
+        &RewardsP3ABrowserTest::GetTestResponse, base::Unretained(this)));
+    rewards_service_->SetLedgerEnvForTesting();
+
+    // Other
+    promotion_->Initialize(browser(), rewards_service_);
+    contribution_->Initialize(browser(), rewards_service_);
+
+    rewards_browsertest_util::SetOnboardingBypassed(browser());
+  }
+
+  void TearDown() override { InProcessBrowserTest::TearDown(); }
+
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    // HTTPS server only serves a valid cert for localhost, so this is needed
+    // to load pages from other hosts without an error
+    command_line->AppendSwitch(switches::kIgnoreCertificateErrors);
+  }
+
+  void GetTestResponse(const std::string& url,
+                       int32_t method,
+                       int* response_status_code,
+                       std::string* response,
+                       base::flat_map<std::string, std::string>* headers) {
+    response_->SetExternalBalance(contribution_->GetExternalBalance());
+    response_->Get(url, method, response_status_code, response);
+  }
+
+  void FetchBalance() {
+    base::RunLoop run_loop;
+    rewards_service_->FetchBalance(base::BindLambdaForTesting(
+        [&](ledger::type::Result result, ledger::type::BalancePtr balance) {
+          run_loop.Quit();
+        }));
+    run_loop.Run();
+  }
+
+  void WaitForRewardsInitialization() {
```

Comment:
I'm having a hard time understanding how this waits for rewards initialization. It just checks once and then runs an empty RunLoop which will just finish immediately. I'm seeing failures when running this in Debug mode and I believe it's because there is a race condition here and we're not actually waiting for rewards to be initialized

Reply:
@brave/rewards-client

Reply:
I think this is doing the right thing in this specific case, but I wouldn't doubt that there may be some other issues in this test (or the Rewards tests in general). But I think the call to `Run` here will block until the `OnRewardsInitialized` observer fires and calls `Quit` on this run loop. What may have caused confusion is that this PR did not include that call to `Quit`; that mistake was caught in a follow-up PR. We could clean this up to use the nicer `Watcher` paradigm from upstream, which would also help to clarify the intent. Apologies though if I'm missing a deeper problem.